### PR TITLE
Feature/GT-128/add static modbus tab to interface

### PR DESCRIPTION
**Pull request by gordon-epc**
Add the static modbus tab to the PM user interface.

### CODE REVIEW

File: src/epcpm/staticmodbusmodel.py [hunk 1]
```diff
@@ -0,0 +1,1356 @@
...
+            return None
+
+        return node
+
+    remove_old_on_drop = epyqlib.attrsmodel.default_remove_old_on_drop
+    internal_move = epyqlib.attrsmodel.default_internal_move
+    check = epyqlib.attrsmodel.check_just_children
+
+
+def check_block_offsets_and_length(self):
```
Comment (altendky) [inline src/epcpm/staticmodbusmodel.py:485]: I'm not sure what form this will have in the new stuff.

File: src/epcpm/staticmodbusmodel.py [hunk 2]
```diff
@@ -0,0 +1,1356 @@
...
+        # + merge("id", Model)
+        + merge(
+            "parameter_uuid",
+            DataPoint,
+            TableRepeatingBlockReferenceDataPointReference,
+            DataPointBitfield,
+            DataPointBitfieldMember,
+        )
+    ),
+    merge("abbreviation", TableRepeatingBlock),
```
Comment (altendky) [inline src/epcpm/staticmodbusmodel.py:1282]: We should talk about whether we want abbreviations in the new system.

**Review by altendky — COMMENTED** [review]
Looks about how I expected.  Lots of the comments are just things we need to think over, not an exhaustive list for sure.  Piles of table stuff to decide...

File: src/epcpm/tests/test_project.py [hunk 3]
```diff
@@ -71,7 +74,7 @@ def test_model_values():
     for name in model:
         model[name] = name + "_"
 
-    assert tuple(model.values()) == ("parameters_", "can_", "sunspec_")
+    assert tuple(model.values()) == ("parameters_", "can_", "staticmodbus_", "sunspec_")
```
Comment (epcAnker) [inline src/epcpm/tests/test_project.py:77]: doesnt matters much but I think the order should be consistent with the interface tab order

File: src/epcpm/staticmodbusmodel.py [hunk 4]
```diff
@@ -0,0 +1,1356 @@
...
+        selected_item = editor.currentItem()
+        datum = str(selected_item.data(epyqlib.utils.qt.UserRoles.raw))
+        model.setData(index, datum)
+
+
+@graham.schemify(tag="data_point", register=True)
+@epyqlib.attrsmodel.ify()
+@epyqlib.utils.qt.pyqtify()
+@attr.s(hash=False)
+class DataPoint(epyqlib.treenode.TreeNode):
```
[the remark this answers is not in the document]
Comment (gordon-epc) [reply src/epcpm/staticmodbusmodel.py]: I changed anything named `*DataPoint*` to `*FunctionData*` to differentiate from SunSpec code/wording. The choice of name comes from the modbus documentation using the wording "Function Code" and "Data Request". I merged those two to form "Function Data". If another naming convention is better, let me know.

File: src/epcpm/tests/test_project.py (hunk 3 above)
Comment (gordon-epc) [reply src/epcpm/tests/test_project.py]: I reordered references to static modbus in the code in [this commit](https://github.com/altendky/pm/pull/10/commits/7e2fc74e272b327be111df54ccf02fdead1d5481).

File: src/epcpm/project.py [hunk 5]
```diff
@@ -198,9 +220,15 @@ def update_enumeration_roots(self):
         self.sunspec.list_selection_roots["sunspec types"] = sunspec_types_root
         self.sunspec.list_selection_roots["enumerations"] = enumerations_root
 
+        self.staticmodbus.list_selection_roots[
+            "staticmodbus types"
+        ] = sunspec_types_root
```
[the remark this answers is not in the document]
Comment (gordon-epc) [reply src/epcpm/project.py]: Added staticmodbus types in 7627a599f9d0391ffd59bc906762a2c48201d24e.

File: src/epcpm/staticmodbusmodel.py [hunk 6]
```diff
@@ -0,0 +1,1356 @@
...
+        default=None,
+        converter=epyqlib.attrsmodel.to_str_or_none,
+        metadata=graham.create_metadata(
+            field=marshmallow.fields.String(allow_none=True),
+        ),
+    )
+    epyqlib.attrsmodel.attrib(
+        attribute=set,
+        no_column=True,
+    )
```
[the remark this answers is not in the document]
Comment (gordon-epc) [reply src/epcpm/staticmodbusmodel.py]: Took out get/set in f5c17b5777cb81d5343fe866a95a324b8696d22d.

File: src/epcpm/staticmodbusmodel.py [hunk 7]
```diff
@@ -0,0 +1,1356 @@
...
+                    severity=epyqlib.checkresultmodel.ResultSeverity.error,
+                    message="No parameter connected",
+                )
+            )
+        else:
+            root = self.find_root()
+            parameter = root.model.node_from_uuid(self.parameter_uuid)
+
+            if (
+                parameter.tree_parent.name.startswith("Static Modbus")
```
[the remark this answers is not in the document]
Comment (gordon-epc) [reply src/epcpm/staticmodbusmodel.py]: Commented out in 8acdd06f64bf270239f398f132d1924ae155b029

File: src/epcpm/staticmodbusmodel.py (hunk 1 above)
Comment (gordon-epc) [reply src/epcpm/staticmodbusmodel.py]: Commented out in 8acdd06f64bf270239f398f132d1924ae155b029

File: src/epcpm/staticmodbusmodel.py (hunk 2 above)
Comment (gordon-epc) [reply src/epcpm/staticmodbusmodel.py]: Commented out in ba23dc821ff9b6427c221e05870fcbf11ae43ac2

**Review by altendky — APPROVED** [review]

**Pull request merged**